ssl/s3_both.cc

```diff
@@ -686,36 +686,28 @@ class CipherScorer {
 };
 
 const SSL_CIPHER *ssl_choose_tls13_cipher(
-    CBS cipher_suites, bool has_aes_hw, uint16_t version, uint16_t group_id,
-    const STACK_OF(SSL_CIPHER) *tls13_ciphers) {
-  if (CBS_len(&cipher_suites) % 2 != 0) {
-    return nullptr;
-  }
+    const STACK_OF(SSL_CIPHER) *client_cipher_suites, bool has_aes_hw, uint16_t version,
+    uint16_t group_id, const STACK_OF(SSL_CIPHER) *tls13_ciphers) {
 
   const SSL_CIPHER *best = nullptr;
   CipherScorer scorer(has_aes_hw);
   CipherScorer::Score best_score = scorer.MinScore();
 
-  while (CBS_len(&cipher_suites) > 0) {
-    uint16_t cipher_suite;
-    if (!CBS_get_u16(&cipher_suites, &cipher_suite)) {
-      return nullptr;
-    }
-
-    SSL_CIPHER *candidate = nullptr;
+  for (size_t i = 0; i < sk_SSL_CIPHER_num(client_cipher_suites); i++) {
+    const SSL_CIPHER *client_cipher = sk_SSL_CIPHER_value(client_cipher_suites, i);
+    const SSL_CIPHER *candidate = nullptr;
     if (tls13_ciphers != nullptr) {
       // Limit to customized TLS 1.3 ciphers if configured.
-      uint32_t cipher_suite_id = 0x03000000L | cipher_suite;
```

Comment:
What does `0x03000000L | cipher_suite` do? Why don't we need it anymore?

Reply:
Ah, great question. I should have called this out on the PR. The magical value `0x03000000L` is `|`'d onto each wire-protocol ciphersuite ID [by OpenSSL](https://github.com/aws/aws-lc/blob/6a1e61a81bb5a1f77bcbf94a1b451b89d939d59b/ssl/ssl_cipher.cc#L1452-L1454). Previously, this section code was doing direct ID equality comparisons on L711 between wire protocol `cipher_suite_id` and OpenSSL ciphersuite IDs, so they had to do the `| 0x03...L` operation here for that to work.

After this refactor, we obtain the `SSL_CIPHER` struct by wire protocol value using [a function which prepends](https://github.com/aws/aws-lc/blob/6a1e61a81bb5a1f77bcbf94a1b451b89d939d59b/ssl/ssl_cipher.cc#L1443) the magic value for us. We then do comparisons using only `SSL_CIPHER.id`, so we [don't need to worry about the conversion](https://github.com/aws/aws-lc/blob/6a1e61a81bb5a1f77bcbf94a1b451b89d939d59b/ssl/internal.h#L544).

To prove this out, I modified this loop with the following diff:

```
$ git diff
diff --git a/ssl/s3_both.cc b/ssl/s3_both.cc
index d09954daa..4b80bc161 100644
--- a/ssl/s3_both.cc
+++ b/ssl/s3_both.cc
@@ -695,11 +695,17 @@ const SSL_CIPHER *ssl_choose_tls13_cipher(

   for (size_t i = 0; i < sk_SSL_CIPHER_num(client_cipher_suites); i++) {
     const SSL_CIPHER *client_cipher = sk_SSL_CIPHER_value(client_cipher_suites, i);
+    if ((client_cipher->id & 0xff000000) != 0x03000000) {
+        abort();
+    }
     const SSL_CIPHER *candidate = nullptr;
     if (tls13_ciphers != nullptr) {
       // Limit to customized TLS 1.3 ciphers if configured.
       for (size_t j = 0; j < sk_SSL_CIPHER_num(tls13_ciphers); j++) {
         const SSL_CIPHER *cipher = sk_SSL_CIPHER_value(tls13_ciphers, j);
+        if ((client_cipher->id & 0xff000000) != 0x03000000) {
+            abort();
+        }
         if (cipher != nullptr && cipher->id == client_cipher->id) {
           candidate = (SSL_CIPHER *)cipher;
           break;
```

and successfully ran the new test after rebuilding:

```
$ cmake-build.sh && ./build/ssl/ssl_test --gtest_filter='*SSLVersionTest.SSLClientCiphers*'; echo $?
...
[==========] 8 tests from 1 test suite ran. (8 ms total)
[  PASSED  ] 8 tests.
0
```